pallets/stateful-storage/src/lib.rs

```diff
@@ -260,17 +263,94 @@ pub mod pallet {
 
 		#[pallet::call_index(1)]
 		#[pallet::weight(0)]
-		pub fn upsert_page(_origin: OriginFor<T>, payload: Vec<u8>) -> DispatchResult {
+		pub fn upsert_page(
+			origin: OriginFor<T>,
+			state_owner_msa_id: MessageSourceId,
+			schema_id: SchemaId,
+			page_id: PageId,
+			payload: Vec<u8>,
+		) -> DispatchResult {
 			ensure!(
 				payload.len() as u32 <= T::MaxPaginatedPageSizeBytes::get(),
 				Error::<T>::PageExceedsMaxPageSizeBytes
 			);
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			let provider_key = ensure_signed(origin)?;
+			let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
+				.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
+
+			let schema = T::SchemaProvider::get_schema_by_id(schema_id);
```

Comment:
Code golf:
	```
	let schema = T::SchemaProvider::get_schema_by_id(schema_id).amp_err (|_|  Error::<T>::InvalidSchemaId)?;
	```

Doing this will allow you to get rid of unwrap() below and do:
```
ensure!(
				schema.payload_location == PayloadLocation::Paginated,
				Error::<T>::SchemaPayloadLocationMismatch
			);
```

Reply:
But that method doesn't return a `Result`, it returns an `Option`. But you got me thinking--a little checking, and I can use `.ok_or()` instead of `.map_err()`. Thanks.